Major rewrite by Hendrik. (#26)
hannestschofenig/lamps-attestation-freshness: draft-ietf-lamps-attestation-freshness.md

Reviewer: I did not work on the following sections so far and will remove dublicated text with the next update.

Reviewer: Use of epoch marker is out of scope of this documebnt anyhow. I propose deleting this section.

Author: See above. I think we have to say a few words about epochs since the Epoch Marker draft is an adopted RATS working group item and we support the functionlity with this document

Reviewer: May be I missunderstand the Epoch marker concept, but as states above, I feel like they do not fit the request/response patten of this draft and therefore are out of scope.

Reviewer: We move it ti an Issue

    seriesinfo:
      ITU-T Recommendation X.690
informative:
  I-D.ietf-rats-msg-wrap:

>   I-D.richardson-rats-composite-attesters:

Reviewer 2: ```suggestion
  I-D.richardson-rats-composite-attesters:
  I-D.ietf-rats-reference-interaction-models:
```

  nonce of that size. If the RA/CA doesn't need a freshness proof, the nonce
  MUST be an empty or zero-length string.
- expiry: In this OPTIONAL field, the RA/CA can specify the validity period of
  the nonce in seconds as an integer value. The nonce can be used during this
  period; the response therefore needs to be conveyed promptly.
- type: In this OPTIONAL field, the RA/CA can specify the type of the respInfo
  structure. The type in the nonce response message is defined by the type in
  the nonce request message.
- respInfo: If type is set, this OPTIONAL field MUST contain the type-specific
  content requested by the end entity for generating the Evidence. If type is
  not set, respInfo MUST also be omitted.

This document does not further specify the content of the reqInfo and respInfo
structures; those structures must be defined elsewhere. Each definition must
assign an object identifier and specify the exact content of the corresponding
field. The message structure and the reqInfo or respInfo structures can use
different encodings. For example, an ASN.1 message can contain reqInfo or
respInfo encoded as JSON, if necessary.

> 

Reviewer 2: ```suggestion
reqInfo may be used to provide the required information to the Relying Party to route the nonce request to the appropriate Verifier when multiple verifiers are supported. respInfo may be used to convey Generic Information Elements ({{Section 6 of I-D.ietf-rats-reference-interaction-models}}) such as Attesting Environment IDs and Claim Selection.

```

- All map keys are text strings.
- The "nonce" member contains the nonce values conveyed as CBOR byte
  strings are between 8 and 64 bytes in length. A zero-length octet string indicates
  that the RA/CA doesis unable or not willing to deliver the requested nonce.
- The OPTIONAL dotted-decimal-oid "type" member denotes a text string containing
  an object identifier in dotted-decimal notation.
- The OPTIONAL "reqInfo" and "respInfo" members contain type-specific CBOR
  values. They MUST only be included if the corresponding "type" member contains
  an OID. Their CBOR encoding is defined by that OID.

~~~~ cddl
nonce-request = {
  ? "len": nonce-length,
  ? "type": dotted-decimal-oid,
  ? "reqInfo": any
}

nonce-response = {
    "nonce": h'' / nonce,
  ? "expiry": uint,
  ? "type": dotted-decimal-oid,
  ? "respInfo": any
}

nonce-length = 8..64
nonce = bstr .size (8..64)
dotted-decimal-oid = tstr
~~~~

# Use with CMP {#CMP}

The nonce request and nonce response message content is conveyed as ASN.1 content,
see {{ASN.1}}, in a general message (genm) {{Section 5.3.19 of RFC9810}}
and general response (genp) {{Section 5.3.20 of RFC9810}}, respectively.

~~~~

> GenMsg:    {id-it TBD1}, NonceRequest

Reviewer 2: ```suggestion
GenMsg:    {id-it TBD1}, nonceRequestMessage
```
Also updated in ASN.1 Module.

Reviewer: not needed

> GenRep:    {id-it TBD2}, NonceResponse

Reviewer 2: ```suggestion
GenRep:    {id-it TBD2}, nonceResponseMessage
```

Reviewer: not needed

the applicable certificate lifecycle management protocols: CMP, EST, and CMC.

# Architecture and Message Formats {#Architecture}

According to {{I-D.ietf-lamps-csr-attestation}}, a CSR can contain multiple
AttestationStatements and multiple certificates for validating those
AttestationStatements. This document describes how an end entity can use CMP,
EST, or CMC to request a nonce from the RA/CA for a specific type of Evidence
to be included in the CSR. The end entity sends a nonce request message with the
following fields:

- len: In this optional field, the end entity can specify the desired length of
  the requested nonce in bytes as a value between 8 and 64.
- type: In this optional field, the end entity can specify the type of the
  reqInfo structure.
- reqInfo: If type is set, this optional field MUST contain the type-specific
  content that the RA/CA requires to generate respInfo. If type is not set,
  reqInfo MUST also be omitted.

The RA/CA can return the requested nonce in a nonce response message together
with information specific to the generation of the Evidence. The nonce response
message has the following fields:

> - nonce: This field contains the nonce. The length of the nonce SHOULD be as

Reviewer 2: ```suggestion
- nonce: This field contains the nonce. The length of the nonce MUST be as
```
consistent with line 205

Reviewer: Thank you for spotting this. the normative language should be consistent, but I am uncertain if MUST or SHOULD is better.
@hannestschofenig What do you think?

Reviewer: I changed it to SHOULD.

  requested in the len field.
- expiry: In this optional field, the RA/CA can specify the validity period of
  the nonce in seconds as an integer value. The nonce can be used during this
  period; the response therefore needs to be conveyed promptly.
- type: In this optional field, the RA/CA can specify the type of the respInfo
  structure. The decision to populate the respInfo with a type
  field will depend on the specification associated with the OID listed in the
  type field of the nonce request message. If the type field is present in both the

>   nonce request and the nonce response message content, the two values MUST match.

Reviewer 2: ```suggestion
  type field of the nonce request message. 
```
I expect the `reqInfo` and `respInfo` will have different values. So IMO it will be better to use different OIDs (ease the implementation).

Reviewer: That's what I wanted to say. The two Type values must match each other, so the infoResp type must match the infoReq type, but they do not have to be identical.
Do you have a suggestion that better expresses this correlation of both types?

Reviewer: Proposal
OLD
"The decision to populate the respInfo with a type
  field will depend on the specification associated with the OID listed in the
  type field of the nonce request message. If the type field is present in both the
  nonce request and the nonce response message content, the two values MUST match."
NEW
"The type in the nonce response message is defined by the type in the nonce request message."

Reviewer: I will change to the NEW

content in a CMP general message request (genm) {{Section 5.3.19 of RFC9810}}
and general response (genp) {{Section 5.3.20 of RFC9810}}, respectively.

~~~~
GenMsg:    {id-it TBD1}, NonceRequest
GenRep:    {id-it TBD2}, NonceResponse | < absent >

id-it-nonceRequest OBJECT IDENTIFIER ::= { id-it TBD1 }
nonceRequestMessage ::= NonceRequest

id-it-nonceResponse OBJECT IDENTIFIER ::= { id-it TBD2 }
nonceResponseMessage ::= NonceResponse
~~~~

When CMP is transferred over HTTP, the OPTIONAL `<operation>` path segment
defined in {{Section 3.4 of RFC9811}} MAY be used for the nonce request message.

| Operation | Operation path | Details |
| --- | --- | --- |
| Get Attestation Freshness Nonce | `getnonce` | {{CMP}} |

When CMP is transferred over CoAP, the OPTIONAL `<operation>` path segment
defined in {{Section 2.1 of RFC9482}} MAY be used for the nonce request message.

| Operation | Operation path | Details |
| --- | --- | --- |
| Get Attestation Freshness Nonce | `nonce` | {{CMP}} |

If the RA/CA does not want to return a nonce, there are several ways to
indicate this. Which variant fits depends on the circumstances.

> - Return a zero-length octet string in the NonceRequest structure.

Reviewer 2: ```suggestion
- Return a zero-length octet string in the NonceResponse structure.
```
But it conflict with the statement in line 205 i.e, RA/CA MUST provide a nonce of that size.

Reviewer: Right

## EST over HTTPS {#EST-https}

If the nonce request and nonce response message content is transmitted over HTTPS,
the specification in {{RFC7030}} applies.

If the nonce request message was successful, the EST server MUST respond with an HTTP 200
status code and the nonce response message content MUST be encoded as JSON object
{{JSON}}. The EST server MUST NOT respond with an HTTP 200 status code if it
does not return a nonce.

In the event of a possible error, the EST server MUST respond with an HTTP
status code 400 (Bad Request) and MUST omit the nonce response message content. If the
nonce request message has been made correctly, but the EST server is unable or unwilling
to deliver the requested nonce, it MUST respond with an HTTP 503 (Service
Unavailable).

The EST server MAY request HTTP-based client authentication as described in
{{Section 3.2.3 of RFC7030}}.

The following media types MUST be used as the content-type of the POST request
and the response.

| Message type<br/>(per operation) | Media type(s) | Reference |
| --- | --- | --- |
| NonceRequest | N/A (for GET) or<br/>application/est-attestation-freshness+json (for POST) | {{EST-https}} |
| --- | --- | --- |

> | NonceResponse   | application/est-attestation- freshness+json | {{EST-https}} |

Reviewer 2: ```suggestion
| NonceResponse   | application/est-attestation-freshness+json | {{EST-https}} |
```

Reviewer: done

the applicable certificate lifecycle management protocols: CMP, EST, and CMC.

# Architecture and Message Formats {#Architecture}

According to {{I-D.ietf-lamps-csr-attestation}}, a CSR can contain multiple
AttestationStatements and multiple certificates for validating those
AttestationStatements. This document describes how an end entity can use CMP,
EST, or CMC to request a nonce from the RA/CA for a specific type of Evidence
to be included in the CSR. The end entity sends a nonceRequest message with the
following fields:

- len: In this optional field, the end entity can specify the desired length of
  the requested nonce in bytes as a value between 8 and 64.
- type: In this optional field, the end entity can specify the type of the
  reqInfo structure.
- reqInfo: If type is set, this optional field MUST contain the type-specific
  content that the RA/CA requires to generate respInfo. If type is not set,
  reqInfo MUST also be omitted.

The RA/CA can return the requested nonce in a nonceResponse message together
with information specific to the generation of the Evidence. The nonceResponse
message has the following fields:

> - nonce: This field contains the nonce. In the ASN.1 encoding, it is an octet

Reviewer: I would prefer moving the encoding-specific information to the respective subsections. 
Protocol-specific informatgion should be covered by those subsections.

Author: Good idea. I wanted to move all encoding specific text but have obviously missed this one

  string between 8 and 64 bytes. If the RA/CA does not want to or cannot return
  a nonce using the ASN.1 encoding, this field contains a zero-length octet
  string. EST reports such failures using HTTP or CoAP error responses as
  described in {{EST}}.
- expiry: In this optional field, the RA/CA can specify the validity period of
  the nonce in seconds as an integer value. The nonce can be used during this
  period; the response therefore needs to be conveyed promptly.
- type: In this optional field, the RA/CA can specify the type of the respInfo

>   structure. The RA/CA MAY copy the value from the type

Reviewer: I would drop this sentnence. The OIDs specifiying the reqInfo and respInfo structures will only equal, if the defuinen syntax equals.

Author: I am OK with dropping it

  field in the nonceRequest. The decision to populate the respInfo with a type
  field will depend on the specification associated with the OID listed in the
  type field of the nonceRequest. If the type field is present in both the

>   nonceRequest and the nonceResponse, the two values MUST match.

Reviewer: If the OIDs are equal not the values but only the syntax must match. I would drop this detail. The information that reqInfo and respInfo are defined by OIDs should explain everything, including that the OIDs can be equal if the syntax shall be equal.

    |                             |---------------------------->|
    |                             |  nonce                      |
    |                             |<----------------------------|
    |  nonce                      |                             |
    |<----------------------------|                             |
    |  attested CSR               |                             |
    |---------------------------->|                             |
    |                             |  Evidence                   |
    |                             |---------------------------->|
    |                             |  Attestation Result         |
    |                             |<----------------------------|
    |  certificate                |                             |
    |<----------------------------|                             |
    |                             |                             |
~~~~
{: #fig-msgFlow title="Message Flow in Background Check Model"}

The nonceRequest and nonceResponse messages allow the end entity to request
only one nonce and one respInfo structure from the RA/CA. If the end entity

> wants to include multiple Evidence statements in a CSR, it should use the

Reviewer: Should we use normative language in this section?

Author: I think we should ask the RATS for review on this aspect since the attester processing is not really in scope of this draft.

Reviewer: Do you want to do the alignment with RATS before or after submitting a new version?

Reviewer: Paragraph was rephrased

8 and 64 bytes in length.

## JSON Representation {#JSON}

This section defines NonceRequest and NonceResponse as JSON objects for use in
EST {{EST}}.

The JSON structure has the following members:

- The optional "len" and "expiry" members, if present, MUST be unsigned
  integers.
- The "nonce" member MUST be a JSON string containing the unpadded base64url
  encoding, as specified in {{Section 5 of RFC4648}}, of a nonce value between
  8 and 64 bytes in length. Such encodings are between 11 and 86 characters in
  length.
- The optional "type" member, if present, MUST be a text string containing the
  object identifier as a dotted-decimal OID.
- The optional "reqInfo" and "respInfo" members MUST only be included if the
  corresponding "type" member contains an OID. Their contents are defined by
  that OID.

{{fig-example-json}} shows an example:

~~~~
 {
   "len": 32,
   "type": "1.2.3.4.5",
   "reqInfo": { ... }
 }

 {
   "nonce": "MTIzNDU2Nzg5MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTI",
   "expiry": 600,
   "type": "1.2.3.4.5",
   "respInfo": { ... }
 }
~~~~
{: #fig-example-json title="NonceRequest and NonceResponse Examples (JSON)"}

> ## CBOR Representation {#CBOR}

Reviewer: I would restructured this section according to the JSON section.

This section defines NonceRequest and NonceResponse as a CBOR structure for use in
EST-coaps {{EST}}. Map keys are text strings. The structures are defined in
CDDL {{RFC8610}} as follows:

~~~~ cddl
nonce-request = {
  ? "len": nonce-length,
  ? "type": dotted-decimal-oid,
  ? "reqInfo": any
}

nonce-response = {
    "nonce": nonce,
  ? "expiry": uint,
  ? "type": dotted-decimal-oid,
  ? "respInfo": any
}

nonce-length = 8..64
nonce = bstr .size (8..64)
dotted-decimal-oid = tstr
~~~~

Nonce values conveyed as CBOR byte strings are between 8 and 64 bytes in
length.

The dotted-decimal-oid type denotes a text string containing an object
identifier in dotted-decimal notation.

The optional "reqInfo" and "respInfo" members contain type-specific CBOR
values. They MUST only be included if the corresponding "type" member contains
an OID. Their CBOR encoding is defined by that OID.

> # Use with CMP {#CMP}

Reviewer: I will add some text for error reporting.

    "certificate-name": ["aik-1"]
  }
}

HTTP/1.1 200 OK
Content-Type: application/est-attestation-freshness+json

{
  "nonce": "MTIzNDU2Nzg5MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTI",
  "expiry": 600,
  "type": "1.2.3.4.5",
  "respInfo": {
    "certificate-name": "aik-1"
  }
}
~~~

## EST over secure COAP {#EST-coaps}

> If the nonce request and nonce response message content is be transmitted via

Reviewer 2: ```suggestion
If the nonce request and nonce response message content is transmitted via
```

Reviewer: done

unwilling to deliver the requested nonce, it MUST respond with a code 5.03
(Service Unavailable).

The following media type MUST be used for nonce request and nonce response message
content according to {{RFC9148}} for CoAP Content Format and Accept option as the
content-type.

| Message type<br/>(per operation) | Media type | Reference |
| --- | --- | --- |
| nonce request<br/>nonce response message content | application/est-attestation-freshness+cbor | {{EST-coaps}} |

# Use with CMC {#CMC}

The nonce request and nonce response message content is conveyed as ASN.1,
see {{ASN.1}}, as CMC Controls in a Full PKI Requests, see
{{Section 6 of I-D.ietf-lamps-csr-attestation}}. The received nonce can be
used for a CSR to be transferred in a Simple or Full PKI request.

To transfer the controls, the content type id-data SHOULD be used.

~~~~
cmc-nonceReq CMC-CONTROL ::=
    { NonceRequest IDENTIFIED BY id-cmc-nonceReq }
id-cmc-nonceReq OBJECT IDENTIFIER ::= { id-cmc TBD4 }

cmc-nonceResp CMC-CONTROL ::=
    { NonceResponse IDENTIFIED BY id-cmc-nonceResp }
id-cmc-nonceResp OBJECT IDENTIFIER ::= { id-cmc TBD5 }
~~~~

> The following example shows a the nonce request and nonce response message content:

Reviewer 2: ```suggestion
The following example shows the nonce request and nonce response message content:
```

Reviewer: done

           {103, id-cmc-nonceResp, <nonce response>}
~~~

> # Nonce Processing Guidelines {#nonce-guidelines}

Reviewer: This section is not only on nonce prosessing. Would it make more sens to name it "Operational Considerations". I will propose this with the next update.

Author: Operational Considerations sounds good to me!

responsible for validating the Epoch Marker and for applying the replay and
freshness rules associated with the selected Epoch Marker type.

This document specifies the transport of a nonce in certificate lifecycle management
protocols. Defining a general freshness-token abstraction that supports Epoch
Markers as a separate token type is out of scope for this version of the
specification.

When Epoch Markers are used for freshness of Evidence, the requirements outlined for nonce generation in {{nonce-guidelines}} do not apply and implementers MUST follow the procedures in {{I-D.ietf-rats-epoch-markers}}.

> #  IANA Considerations {#iana}

Reviewer: I will update this Section and streamline it a bit.

Author: I added the reference to the Epoch markers because I got a request from Henk about the description. Their document in the RATS group on epoch is actually using the nonce encoding and so we are actually supporting both, regular nonces and epochs.

Reviewer: If we see request/response of Epoch marker as alternative to nonces in scope of this document, we should reconsider the focus on nonces in the title, abstract, and introduction.
My understanding was, that Epoch marker are not requested, but provided by the RR on a regular basis. This does not fit the request/response patten of the protocols used in this draft.

Reviewer: @hannestschofenig  please open an Issue

## CMP HTTP Path Segment

IANA is requested to add the following entry to the "CMP Well-Known URI Path
Segments" registry defined in {{RFC8615}}:

| Path Segment | Description | Reference |
| --- | --- | --- |
| `getnonce` | Get Attestation Freshness Nonce over HTTP | {{CMP}} |

## CMP CoAP Path Segment

IANA is requested to add the following entry to the "CMP Well-Known URI Path
Segments" registry defined in {{RFC8615}}:

| Path Segment | Description | Reference |
| --- | --- | --- |
| `nonce` | Get Attestation Freshness Nonce over CoAP | {{CMP}} |

> ## EST Operation Path Segment

Reviewer: Currently none of the /est operational path segments are registered at IANA. therefor I would also not ask for registration of /nonce.

IANA is requested to add the following EST operation path segment under the
existing "/.well-known/est" URI path-prefix defined in {{RFC7030}}:

| Operation | Operation path | Details |
| --- | --- | --- |
| Retrieval of a nonce | `/nonce` | {{EST}} |

> ## Well-Known URI Reference Update

Reviewer: Is this addition needed? Currently the EST and EST-COAP base-RFCs are listed.

identified by the `AttestationStatement` type. The returned Evidence is encapsulated
in the `AttestationStatement` within the `AttestationBundle` carried in the CSR, as
defined in {{I-D.ietf-lamps-csr-attestation}}. The software generating the CSR
treats the attestation statement payload as an opaque blob and does not
interpret its format. It is crucial to note that the nonce is included in the
Evidence, either implicitly or explicitly, and MUST NOT be conveyed in CSR
structures outside of the attestation payload.

The freshness established by the nonce applies to the Evidence as
bound to the nonce used by the attestation technology. This
specification does not assign independent freshness semantics to other claims
contained in the Evidence. The interpretation of any other claim, including
whether it represents current state, historical state, configuration state, or
measurement results, is defined by the attestation technology, the appraisal
policy, and the Verifier or Relying Party processing rules.

The processing of CSRs containing attestation statements is detailed in
{{I-D.ietf-lamps-csr-attestation}}. Importantly, certificates issued based on

> this process do not contain the nonce, as specified in

Reviewer 2: ```suggestion
{{I-D.ietf-lamps-csr-attestation}}. Importantly, certificates issued based on
this process should not contain the Attestation, as specified in
```

Reviewer: I could not find any such statement in csr-attestation and propose to remove the paragraphe completely.
@hannestschofenig what do you think?

Reviewer: OK, will remove it

{{RFC9334}} defines the IETF remote attestation architecture and
extensively discusses nonce-based freshness.

Specific attestation technology specification such as {{RFC9711}} and

> {RFC9783}} offers guidance on replay protection using nonces. This

Reviewer 2: ```suggestion
{{RFC9783}} offers guidance on replay protection using nonces. This
```

Reviewer: Done

{{RFC9334}} defines the IETF remote attestation architecture and
extensively discusses nonce-based freshness.

> {{Section 8.4 of RFC9711}} specifies requirements for replay protection and

Reviewer: I think this paragraph does not belong tothis draft. The next paragraph explains why.

Author: I added it as an example supporting the statement about the nonce length.

Reviewer: What is you proposal?

Reviewer: leave it out

    len    INTEGER (8..64) OPTIONAL,
    -- indicates the required length of the requested nonce
    type   ATTESTATION-NONCE-REQUEST.&id(
              {AttestationNonceRequestSet}) OPTIONAL,
    -- identifies the nonce-request syntax for the selected
    --   attestation statement type
    reqInfo ATTESTATION-NONCE-REQUEST.&Type(
              {AttestationNonceRequestSet}{@type}) OPTIONAL
    -- contains type-specific nonce-request information
 }

 NonceResponse ::= SEQUENCE {
    nonce  OCTET STRING (SIZE(0 | 8..64)),
    -- contains the nonce of length len
    -- a zero-length OCTET String indicate that no freshness
    --   proof is required
    expiry INTEGER OPTIONAL,
    -- indicates how long in seconds the nonce issuer considers
    --   the nonce valid
    type   ATTESTATION-NONCE-RESPONSE.&id(
              {AttestationNonceResponseSet}) OPTIONAL,
    -- identifies the nonce-response syntax for the selected
    --   attestation statement type
    respInfo ATTESTATION-NONCE-RESPONSE.&Type(
               {AttestationNonceResponseSet}{@type}) OPTIONAL
    -- contains type-specific nonce-response information
 }

 id-it-nonceRequest OBJECT IDENTIFIER ::= { id-it TBD1 }

>     nonceReq ::= NonceRequest

Reviewer 2: ```suggestion
    nonceRequestMessage ::= NonceRequest
```

Reviewer: I will change it

 id-it-nonceResponse OBJECT IDENTIFIER ::= { id-it TBD2 }

>     nonceResp ::= NonceResponse

Reviewer 2: ```suggestion
    nonceResponseMessage ::= NonceResponse
```

Reviewer: I will change it

    -- contains type-specific nonce-response information
 }

 id-it-nonceResponse OBJECT IDENTIFIER ::= { id-it TBD2 }

>     nonceReq ::= NonceRequest

Reviewer 2: ```suggestion
    nonceResp ::= NonceResponse
```

Reviewer: Done

 id-it-nonceRequest OBJECT IDENTIFIER ::= { id-it TBD1 }

>     nonceResp ::= NonceResponse

Reviewer 2: ```suggestion
    nonceReq ::= NonceRequest
```

Reviewer: Done
